java/arcs/core/util/Scheduler.kt

```diff
@@ -83,23 +84,33 @@ class Scheduler(
                 pausedChannel.asFlow().filter { !it }.first()
                 launches.incrementAndGet()
 
-                idlenessChannel.send(false)
                 try {
                     withTimeout(agendaProcessingTimeoutMs) { executeAgenda(agenda) }
                 } finally {
-                    idlenessChannel.send(true)
+                    val agendasLeft = agendasInFlight.getAndDecrement()
+                    if (agendasLeft == 1) {
+                        idlenessChannel.send(true)
+                    }
                 }
             }
             .launchIn(scope)
     }
 
     /** Schedule a single [Task] to be run as part of the next agenda. */
     fun schedule(task: Task) {
+        val agendasInFlight = agendasInFlight.getAndIncrement()
+        if (agendasInFlight == 0) {
+            idlenessChannel.offer(false)
```

Comment:
Just because `offer `is non-suspending. I could use it in both places actually, since the channel is a ConflatedBroadcastChannel, and the boolean wouldn't be dropped (offer never returns false for ConflatedBroadcastChannel)